### PR TITLE
Add prefill packing for offline inference

### PR DESCRIPTION
**Pull request by sixiang-google**
# Description

Concatenates the prefills of several prompts for higher MFU in prefills. For now, only support bucketized concatenation for fixed padded lengths (mainly 128, 256 and 512). 

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed.

### CODE REVIEW

File: MaxText/maxengine.py [hunk 1]
```diff
@@ -468,6 +565,123 @@ def copy(path, partial_cache, full_cache, annotations):
...
+  def insert_partial(
+      self,
+      prefix: Prefix,
+      decode_state: DecodeState,
+      slot: int,
+      *,
+      start_idx: int,
+      seq_len: int,
+  ) -> DecodeState:
+    """Insert into KV cache"""
```
Comment (vipannalla) [inline MaxText/maxengine.py:586]: Please update doc to indicate in what scenario this method should be used.

File: MaxText/inference_mlperf/offline_inference.py [hunk 2]
```diff
@@ -125,20 +184,73 @@ def batch_inference_with_callback(
...
+        prefill_result = []
+        prefill_fn = self._prefill_insert
+        if (cached := self._cached_pref.get(prefill_len)) is not None:
+          prefill_fn = cached
+        for slot, row in prefill_bucket:
+          first_token, self.decode_state = prefill_fn(
+              self.params, tokens=row.tokens, slot=slot, true_length=row.true_length, decode_state=self.decode_state
+          )
+          prefill_result.append((first_token, slot, row))
+        return prefill_result
```
Comment (vipannalla) [inline MaxText/inference_mlperf/offline_inference.py:202]: Have you tried to batch prefills and pad to closet power-of-2? If not, maybe add a TODO to try in future

File: MaxText/inference_mlperf/offline_inference.py [hunk 3]
```diff
@@ -125,20 +184,73 @@ def batch_inference_with_callback(
...
 
-      prefill_fn = self._prefill_insert
-      if (cached := self._cached_pref.get(len(tokens))) is not None:
-        prefill_fn = cached
-
-      first_token, self.decode_state = prefill_fn(
-          self.params, tokens=tokens, slot=slot, true_length=true_length, decode_state=self.decode_state
-      )
-      return first_token
+        return prefill_result
```
Comment (vipannalla) [inline MaxText/inference_mlperf/offline_inference.py:253]: You can move this packing code to a separate util (or to max_utils) for re-use.

**Review by vipannalla — COMMENTED** [review]
Looks great, added few minor comments. Can you also add unittests for packing and unpacking code?

File: MaxText/inference_mlperf/offline_inference.py [hunk 4]
```diff
@@ -224,16 +339,36 @@ def detokenize():
...
+      log.info(f"Total num prefill: {num_prefill}")
       slot = empty_slots.pop()
-      first_token = prefill(slot, row.tokens, row.true_length)
-      self.detokenize_backlog.put((first_token, True, row.id, slot), block=True)
-
+      # directly prefill prompts with 64 or less tokens
+      if num_tokens == 64:
+        first_token, slot, row = prefill([(slot, row)], 64)[0]
+        self.detokenize_backlog.put((first_token, True, row.id, slot), block=True)
+        continue
```
[the remark this answers is not in the document]
Comment (sixiang-google) [reply MaxText/inference_mlperf/offline_inference.py:353]: The slots are set to empty once eos is received but filled when batch prefill is performed; therefore collecting a bin of padded 64 tokens prompts would result in <16 slots empty running while waiting for the bucket to fill. Also prompts with <64 tokens are only a small fraction. I experimented and this loss is more than the time saved with packing.

File: MaxText/inference_mlperf/offline_inference.py (hunk 2 above)
Comment (sixiang-google) [reply MaxText/inference_mlperf/offline_inference.py:202]: The current batch size is configured for largest cache size of 1024

File: MaxText/maxengine.py [hunk 5]
```diff
@@ -310,6 +311,102 @@ def prefill(
...
+          length_idx=(2, 3),
+          samples_per_slot=1,
+      )
+      return {
+          "logits": selected_logits,
+          "cache": cache,
+          "next_pos": next_pos,
+          "generated_tokens": generated_tokens,
+          "tokens": first_generated_token,
+      }, result
```
[the remark this answers is not in the document]
Comment (sixiang-google) [reply MaxText/maxengine.py:400]: I will refactor this in the next PR; the signatures would be totally different if we want to jit the entire prefill packed function together.

File: MaxText/inference_mlperf/offline_inference.py (hunk 3 above)
Comment (sixiang-google) [reply MaxText/inference_mlperf/offline_inference.py:253]: Since we will do some other changes (remove padding, etc. ) in the close future, I would like to keep this over here for flexibility

File: MaxText/inference_mlperf/offline_inference.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (vipannalla) [reply MaxText/inference_mlperf/offline_inference.py:353]: Sounds good, can you add these details as code comment above the if statement?

File: MaxText/maxengine.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (vipannalla) [reply MaxText/maxengine.py:400]: Sounds good, just add a TODO comment to track.

[the remark this answers is not in the document]
Comment (sixiang-google) [reply MaxText/maxengine.py:409]: Its ready. Will commit once this goes through

File: MaxText/inference_mlperf/offline_inference.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sixiang-google) [reply MaxText/inference_mlperf/offline_inference.py:352]: Added

File: MaxText/inference_mlperf/offline_inference.py [hunk 6]
```diff
@@ -224,16 +339,36 @@ def detokenize():
...
+        for first_token, slot, row in prefill_results:
+          log.debug(f"Put row of len {row.tokens.shape[0]} true length {row.true_length} slot {slot} to detokenize backlog")
+          self.detokenize_backlog.put((first_token, True, row.id, slot), block=True)
+        self.prefill_buckets[num_tokens] = []
+
+    for num_tokens in self.prefill_buckets.keys():
+      prefill_results = prefill(self.prefill_buckets[num_tokens], num_tokens)
+      for first_token, slot, row in prefill_results:
+        log.debug(f"Put row of len {row.tokens.shape[0]} true length {row.true_length} slot {slot} to detokenize backlog")
+        self.detokenize_backlog.put((first_token, True, row.id, slot), block=True)
```
[the remark this answers is not in the document]
Comment (sixiang-google) [reply MaxText/inference_mlperf/offline_inference.py:367]: Added

File: MaxText/maxengine.py [hunk 7]
```diff
@@ -310,6 +311,102 @@ def prefill(
...
+      sampler: Optional[Callable[[Any], Any]] = None,  # pylint: disable=unused-argument
+      rng: Optional[jax.random.PRNGKey] = None,
+  ) -> Tuple[PackedPrefix, engine_api.ResultTokens]:
+    """Computes a kv-cache for a new generate request.
+
+    Args:
+      params: Scalar multiplier.
+      existing_prefix: If provided, represents a prefix that has already been
+        processed by the underlying model.
+      padded_tokens: Logically appended tokens to any existing prefix, this is
```
[the remark this answers is not in the document]
Comment (sixiang-google) [reply MaxText/maxengine.py:338]: Done

File: MaxText/maxengine.py [hunk 8]
```diff
@@ -310,6 +311,102 @@ def prefill(
...
+    if existing_prefix:
+      raise ValueError("We don't know what to do with existing_prefix")
+
+    if rng is None:
+      rng = jax.random.PRNGKey(0)
+    input_tokens = jnp.expand_dims(padded_tokens, 0)  # [BATCH, SEQUENCE]
+    decoder_positions = jnp.expand_dims(decoder_positions, 0)
+    decoder_segment_ids = jnp.expand_dims(decoder_segment_ids, 0)
+    rng, new_rng = jax.random.split(rng)
+    with self._mesh, nn_partitioning.axis_rules(self.config.logical_axis_rules):
```
[the remark this answers is not in the document]
Comment (sixiang-google) [reply MaxText/maxengine.py:359]: Added

File: MaxText/maxengine.py (hunk 1 above)
Comment (sixiang-google) [reply MaxText/maxengine.py]: Added

File: MaxText/layers/attentions.py [hunk 9]
```diff
@@ -177,9 +177,14 @@ def generate_attention_mask(self, query, key, decoder_segment_ids: Array | None,
...
       mask = decoder_segment_ids[:, None, None, None, :] == common_types.DECODING_ACTIVE_SEQUENCE_INDICATOR
-    elif decoder_segment_ids is not None:
+    elif decoder_segment_ids is not None and model_mode == common_types.MODEL_MODE_PREFILL:
       mask = decoder_segment_ids[:, :, None] == decoder_segment_ids[:, None, :]
       mask = mask[:, None, None, :, :]
+      zero_segment_ids = decoder_segment_ids == 0
+      zero_mask = jnp.outer(zero_segment_ids[:, :, None], zero_segment_ids[:, None, :])
+      zero_mask = jnp.expand_dims(zero_mask, 0)
+      zero_mask = jnp.logical_not(zero_mask)[:, None, None, :, :]
+      mask = jnp.logical_and(mask, zero_mask)
```
Comment (vipannalla) [inline MaxText/layers/attentions.py:187]: Can you add a comment with example on what this code is doing for future readers?

File: MaxText/layers/attentions.py [hunk 10]
```diff
@@ -177,9 +177,14 @@ def generate_attention_mask(self, query, key, decoder_segment_ids: Array | None,
     mask = None
     if model_mode == common_types.MODEL_MODE_AUTOREGRESSIVE:
       mask = decoder_segment_ids[:, None, None, None, :] == common_types.DECODING_ACTIVE_SEQUENCE_INDICATOR
-    elif decoder_segment_ids is not None:
+    elif decoder_segment_ids is not None and model_mode == common_types.MODEL_MODE_PREFILL:
```
Comment (vipannalla) [inline MaxText/layers/attentions.py:180]: This may break training code since there is not else block to handle `common_types.MODEL_MODE_TRAIN`.

Comment (sixiang-google) [reply MaxText/layers/attentions.py]: Fixed

File: MaxText/layers/attentions.py (hunk 9 above)
Comment (sixiang-google) [reply MaxText/layers/attentions.py]: Added

**Review by vipannalla — APPROVED** [review]
Looks good.
